### PR TITLE
Aks aad

### PR DESCRIPTION
**Pull request by LochanRn**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by LochanRn** [comment]
@jzhoucliqr Please review this PR :)

File: exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go [hunk 1]
```diff
@@ -291,3 +296,147 @@ func (r *AzureManagedControlPlane) validateSSHKey() error {
...
+
+	if old != nil && new == nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("Spec", "AADProfile"),
+				new,
+				"field cannot be nil, cannot disable AADProfile"))
+	}
+
+	if old == nil && new != nil {
```
Comment (jzhoucliqr) [inline exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go:362]: does this section mean for legacy you can't enable aad after cluster is created?

File: exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go [hunk 2]
```diff
@@ -291,3 +296,147 @@ func (r *AzureManagedControlPlane) validateSSHKey() error {
 
 	return nil
 }
+
+func (r *AzureManagedControlPlane) validateAadProfile() error {
```
Comment (jzhoucliqr) [inline exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go:300]: I see lots of immutable here, so overall, which field is mutable? what kinds of update is allowed?

File: exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go (hunk 1 above)
Comment (LochanRn) [reply exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go:362]: yes correct

File: exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go (hunk 2 above)
Comment (LochanRn) [reply exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go:300]: managed can be mutated only to true before and after cluster creation
adminGroupIds can be mutated, but must have minimum one if managed is set to true.

other 4 fields are w.r.t aad and can be disabled only while moving from legacy to managed aks aad support.

File: azure/converters/rules.go [hunk 3]
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 package converters
 
 import (
-	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
```
[the remark this answers is not in the document]
Comment (LochanRn) [reply azure/converters/rules.go:20]: It is a different service and capz has different services using different versions

File: exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go [hunk 4]
```diff
@@ -291,3 +296,100 @@ func (r *AzureManagedControlPlane) validateSSHKey() error {
...
+				allErrs = append(allErrs,
+					field.Invalid(
+						field.NewPath("Spec", "AADProfile", "AdminGroupObjectIDs"),
+						new.ManagedAAD.AdminGroupObjectIDs,
+						"need atleast one AdminGroupObjectID"))
+			}
+		}
+
+		if new.ManagedAAD == nil {
+			if new.LegacyAAD != old.LegacyAAD {
```
Comment (jzhoucliqr) [inline exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go:385]: need to compare each of the fields, maybe deepEqual?

Comment (LochanRn) [reply exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go:385]: yeah okay, makes sense 👍

File: exp/api/v1alpha4/azuremanagedcontrolplane_types.go [hunk 5]
```diff
@@ -86,6 +86,53 @@ type AzureManagedControlPlaneSpec struct {
...
+	LegacyAAD *LegacyAAD `json:"legacyAadProfile,omitempty"`
+}
+
+// ManagedAAD - AAD integration is managed by AKS.
+type ManagedAAD struct {
+	// Managed - Whether to enable managed AAD.
+	// +kubebuilder:validation:Required
+	Managed *bool `json:"managed"`
+
+	// DEPRECATED: use ManagedAAD instead
```
Comment (jzhoucliqr) [inline exp/api/v1alpha4/azuremanagedcontrolplane_types.go:113]: remove

File: exp/api/v1alpha3/azuremanagedcontrolplane_types.go [hunk 6]
```diff
@@ -81,6 +81,53 @@ type AzureManagedControlPlaneSpec struct {
...
+	LegacyAAD *LegacyAAD `json:"legacyAadProfile,omitempty"`
+}
+
+// ManagedAAD - AAD integration is managed by AKS.
+type ManagedAAD struct {
+	// Managed - Whether to enable managed AAD.
+	// +kubebuilder:validation:Required
+	Managed *bool `json:"managed"`
+
+	// DEPRECATED: use ManagedAAD instead
```
Comment (jzhoucliqr) [inline exp/api/v1alpha3/azuremanagedcontrolplane_types.go:108]: remove

**Comment by jzhoucliqr** [comment]
looks good! Thanks @LochanRn !

**Comment by LochanRn** [comment]
so shall i merge this branch and raise a PR to the upstream ?

**Pull request merged**